### ISSUE TITLE
`Guarantee failed: [inserted]` in limit changefeeds

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#6093]
Reported by @analytik in https://github.com/rethinkdb/rethinkdb/issues/5646#issuecomment-244734975 : 

```
Mon Sep  5 08:12:13 UTC 2016 Starting RethinkDB db1 rethinkdb1-6sigm on 10.244.78.38 (rethinkdb1-6sigm)
Running rethinkdb 2.3.5~0jessie (GCC 4.9.2)...
Running on Linux 4.7.1-coreos x86_64
Loading data from directory /data/rethinkdb_data
Listening for intracluster connections on port 29015
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on cluster addresses: 127.0.0.1, 10.244.78.38, ::1, fe80::42:aff:fef4:4e26%78
Listening on driver addresses: 127.0.0.1, 10.244.78.38, ::1, fe80::42:aff:fef4:4e26%78
Listening on http addresses: 127.0.0.1, 10.244.78.38, ::1, fe80::42:aff:fef4:4e26%78
Server ready, "rethinkdb1_pigct_wps" 93feaf90-0fed-4688-a839-96d0f301c312
Connected to proxy proxy-be642938-feeb-4ce2-9ffd-cde7c574e4a7
Version: rethinkdb 2.3.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1145:
error: Guarantee failed: [inserted]
error: Backtrace:
Version: rethinkdb 2.3.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1145:
error: Guarantee failed: [inserted]
error: Backtrace:
error: Mon Sep  5 11:31:09 2016

       1 [0xae7500]: backtrace_t::backtrace_t() at ??:?
       2 [0xae7879]: format_backtrace(bool) at ??:?
       3 [0xd9f6c3]: report_fatal_error(char const*, int, char const*, ...) at ??:?
       4 [0x8e37b5]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at ??:?
       5 [0x7f25a2]: /usr/bin/rethinkdb() [0x7f25a2] at 0x7f25a2 ()
       6 [0x8dec0d]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at ??:?
       7 [0x7f3ce5]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at ??:?
       8 [0x7fcbd9]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at ??:?
       9 [0x9f2c47]: coro_t::run() at ??:?
error: Exiting.
/run_dev.sh: line 10:     8 Trace/breakpoint trap   (core dumped) /usr/bin/rethinkdb --bind all --canonical-address rethinkdb1 --cache-size 256 --cluster-reconnect-timeout 30
```

**Comment by danielmewes** [comment]
@analytik I don't think I've seen this issue before.

It must be related to some open `.orderBy.limit` changefeeds. Do you have information on which precise changefeed queries of that type are open on the cluster during the import?

Can you send us a copy of the data dump that you've been getting this crash with?

**Comment by analytik** [comment]
After a bit of sifting, I managed to find one table and one query that caused this. Using rethinkdbdash@2.3.21 -

```
r.db("wallet").table("currency_rates").orderBy({
    index: r.desc("created")
}).limit(1).changes({
    includeInitial: true
})
```

Just to confirm, I'm pasting the fresh error, just in case it's slightly different.
```
Fri Sep  9 12:03:05 UTC 2016 Starting RethinkDB db1 rethinkdb1-tzwpn on 10.244.96.10 (rethinkdb1-tzwpn)
Running rethinkdb 2.3.5~0jessie (GCC 4.9.2)...
Running on Linux 4.7.1-coreos x86_64
Loading data from directory /data/rethinkdb_data
Listening for intracluster connections on port 29015
Listening for client driver connections on port 28015
Listening for administrative HTTP connections on port 8080
Listening on cluster addresses: 127.0.0.1, 10.244.96.10, ::1, fe80::42:aff:fef4:600a%107
Listening on driver addresses: 127.0.0.1, 10.244.96.10, ::1, fe80::42:aff:fef4:600a%107
Listening on http addresses: 127.0.0.1, 10.244.96.10, ::1, fe80::42:aff:fef4:600a%107
Server ready, "rethinkdb1_g0gba_njk" 801fa476-fc6f-4ff0-9416-dcb9aefd71e5
Connected to proxy proxy-630cab81-07d3-454b-a54d-5b7ecb545c7d
Version: rethinkdb 2.3.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1145:
error: Guarantee failed: [inserted]
error: Backtrace:
Version: rethinkdb 2.3.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1145:
error: Guarantee failed: [inserted]
error: Backtrace:
error: Fri Sep  9 12:05:14 2016

       1 [0xae7500]: backtrace_t::backtrace_t() at ??:?
       2 [0xae7879]: format_backtrace(bool) at ??:?
       3 [0xd9f6c3]: report_fatal_error(char const*, int, char const*, ...) at ??:?
       4 [0x8e37b5]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at ??:?
       5 [0x7f25a2]: /usr/bin/rethinkdb() [0x7f25a2] at 0x7f25a2 ()
       6 [0x8dec0d]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at ??:?
       7 [0x7f3ce5]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at ??:?
       8 [0x7fcbd9]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at ??:?
       9 [0x9f2c47]: coro_t::run() at ??:?
error: Exiting.
/run_dev.sh: line 10:     9 Trace/breakpoint trap   (core dumped) /usr/bin/rethinkdb --bind all --canonical-address rethinkdb1 --cache-size 256 --cluster-reconnect-timeout 30
```

Here's the minimum-size dump that I could still reliably reproduce the issue with:

[rethinkdb_stage_dump_tiny.tar.gz](https://github.com/rethinkdb/rethinkdb/files/463910/rethinkdb_stage_dump_tiny.tar.gz)

**Comment by danielmewes** [comment]
@analytik Thanks, that's super useful!

**Comment by danielmewes** [comment]
Able to reproduce with above query and data set. I had to import the data set once, then run the query, and then import it a second time with the `--force` option to `rethinkdb restore`.

**Comment by danielmewes** [comment]
I think this has something to do with the `restore` utility recreating the secondary index that the changefeed is on. A work-around is running `rethinkdb restore` with the `--no-secondary-indexes` option, and then recreating the secondary indexes manually if they don't exist yet.

**Comment by danielmewes** [comment]
(I *think* we somehow end up with a changefeed on an index that's still under construction, and that  then isn't handled correctly)

**Comment by danielmewes** [comment]
Oh, I think I'm seeing the issue. Limit changefeed managers are bound to index names rather than index UUIDs. During the import process, we get two different indexes that happen to share the same name. Will verify this now...

**Comment by danielmewes** [comment]
I think there's a deeper problem here in that streams internally only know the name of a secondary index, but not its UUID.

It looks like it's possible with any streamed query that an index could be replaced by a different one with the same name in between two query batches, and the query would silently start to use the new index in the middle of streaming results (or maybe it crash the server?).

In the case of changefeeds, this problem is particularly severe, but I think it's not limited to them.

This might be non-trivial to fix, because the server that is processing the query currently has no way of locally resolving an index name to its UUID.

**Comment by danielmewes** [comment]
I have a fix for this particular case in branch daniel_6093 and code review 3750.

There is some evidence that there are some remaining ways to hit this crash in different circumstances (possibly when resharding a table).

**Comment by danielmewes** [comment]
This is now fixed in branch `next` via 3136e81 and `v2.3.x` via 5f82c3a.

It turns out that this also fixes another case we were concerned about for changefeeds.

There could still be some issues with other changefeed types when indexes get recreated while changefeeds are open. This will require additional testing.

**Comment by danielmewes** [comment]
@analytik To explain: running `rethinkdb restore --force` drops existing indexes and then recreates new ones under (potentially) the same names. The code in the `.orderBy.limit` changefeed implementation uses index names to reference indexes. Thus they didn't detect when the index was replaced by a new index of the same name while having the changefeed open. The patch adds a check for this, and the changefeeds now get closed when the index is changed.

**Comment by analytik** [comment]
Perfect, thank you so much @danielmewes. Luckily, we don't run `restore --force` in production, so this was only an issue in development environment, where dropped connections during a full DB restore are expected.

**Comment by analytik** [comment]
@danielmewes this happened last night in production, and we sure weren't running rethinkdb restore there. Can you please check if it's related, or should I post it as a new issue?

It's possible that one of the node.js app servers got restarted and tried to make a db migration and create indexes, but all indexes should exist at this point, and the attempted db migration only does something like `table.indexCreate()` and then ignores all `ReqlOpFailedError`.

```
Version: rethinkdb 2.3.5~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1145:
error: Guarantee failed: [inserted]
error: Backtrace:
error: Tue Sep 13 20:39:39 2016

       1 [0xae7500]: backtrace_t::backtrace_t() at ??:?
       2 [0xae7879]: format_backtrace(bool) at ??:?
       3 [0xd9f6c3]: report_fatal_error(char const*, int, char const*, ...) at ??:?
       4 [0x8e37b5]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at ??:?
       5 [0x7f25a2]: /usr/bin/rethinkdb() [0x7f25a2] at 0x7f25a2 ()
       6 [0x8dec0d]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at ??:?
       7 [0x7f3ce5]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at ??:?
       8 [0x7fcbd9]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at ??:?
       9 [0x9f2c47]: coro_t::run() at ??:?
error: Exiting.
```

**Comment by AtnNn** [comment]
A fix for this issue has been released in [RethinkDB 2.3.6](https://rethinkdb.com/blog/2.3.6-release/)

**Comment by analytik** [comment]
Oops, sorry for double posting here and in #6309 - we still have crashes probably related to this:

```
Running on Linux 4.9.16-coreos-r1 x86_64
Version: rethinkdb 2.3.6~0jessie (GCC 4.9.2)
error: Error in src/rdb_protocol/changefeed.cc at line 1164:
error: Guarantee failed: [inserted]
error: Backtrace:
error: Tue Sep  5 16:59:39 2017

       1 [0x7d6e20]: backtrace_t::backtrace_t() at 0x7d6e20 (/usr/bin/rethinkdb)
       2 [0x7d7199]: format_backtrace(bool) at 0x7d7199 (/usr/bin/rethinkdb)
       3 [0x72b053]: report_fatal_error(char const*, int, char const*, ...) at 0x72b053 (/usr/bin/rethinkdb)
       4 [0x9cf735]: ql::changefeed::limit_manager_t::commit(rwlock_in_line_t*, boost::variant<ql::changefeed::primary_ref_t, ql::changefeed::sindex_ref_t, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_, boost::detail::variant::void_> const&) at 0x9cf735 (/usr/bin/rethinkdb)
       5 [0x99d482]: /usr/bin/rethinkdb() [0x99d482] at 0x99d482 ()
       6 [0x9ca95c]: ql::changefeed::server_t::foreach_limit(boost::optional<std::string> const&, boost::optional<uuid_u> const&, store_key_t const*, std::function<void (rwlock_in_line_t*, rwlock_in_line_t*, rwlock_in_line_t*, ql::changefeed::limit_manager_t*)>, auto_drainer_t::lock_t const&) at 0x9ca95c (/usr/bin/rethinkdb)
       7 [0x99ec17]: rdb_update_single_sindex(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*) at 0x99ec17 (/usr/bin/rethinkdb)
       8 [0x9a7a09]: callable_action_instance_t<std::_Bind<void (*(store_t*, store_t::sindex_access_t*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*))(store_t*, store_t::sindex_access_t const*, deletion_context_t const*, rdb_modification_report_t const*, unsigned long*, auto_drainer_t::lock_t, cond_t*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*, std::vector<std::pair<ql::datum_t, std::string>, std::allocator<std::pair<ql::datum_t, std::string> > >*)> >::run_action() at 0x9a7a09 (/usr/bin/rethinkdb)
       9 [0x815857]: coro_t::run() at 0x815857 (/usr/bin/rethinkdb)
error: Exiting.
```
